src/mongo_logger.ts

```diff
@@ -0,0 +1,144 @@
+import * as fs from 'fs';
+import type { Writable } from 'stream';
+
+import { getUint } from './connection_string';
+
+/** @internal */
+export const SeverityLevel = Object.freeze({
+  EMERGENCY: 'emergency',
+  ALERT: 'alert',
+  CRITICAL: 'critical',
+  ERROR: 'error',
+  WARNING: 'warn',
+  NOTICE: 'notice',
+  INFORMATIONAL: 'info',
+  DEBUG: 'debug',
+  TRACE: 'trace',
+  OFF: 'off'
+} as const);
+
+/** @internal */
+export type SeverityLevel = typeof SeverityLevel[keyof typeof SeverityLevel];
+
+/** @returns one of SeverityLevel or null if passsed severity is not a valid SeverityLevel */
+function toValidSeverity(severity?: string): SeverityLevel | null {
+  const validSeverities: string[] = Object.values(SeverityLevel);
+  const lowerSeverity = severity?.toLowerCase();
+
+  if (lowerSeverity != null && validSeverities.includes(lowerSeverity)) {
+    return lowerSeverity as SeverityLevel;
+  }
+
+  return null;
+}
+
+/** @internal */
+export const MongoLoggableComponent = Object.freeze({
+  COMMAND: 'command',
+  TOPOLOGY: 'topology',
+  SERVER_SELECTION: 'serverSelection',
+  CONNECTION: 'connection'
+} as const);
+
+/** @internal */
+export type MongoLoggableComponent =
+  typeof MongoLoggableComponent[keyof typeof MongoLoggableComponent];
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  mongodbLogPath?: string | Writable;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  /** Severity level for command component */
+  command: SeverityLevel;
+  /** Severity level for SDAM */
+  topology: SeverityLevel;
+  /** Severity level for server selection component */
+  serverSelection: SeverityLevel;
+  /** Severity level for CMAP */
+  connection: SeverityLevel;
+  /** Default severity level to be if any of the above are unset */
+  defaultSeverity: SeverityLevel;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. Must be 'stderr', 'stdout', a file path, or a Writable. Defaults to 'stderr'. */
+  logPath: string | Writable;
+}
+
+/** @internal */
+export class MongoLogger {
+  componentSeverities: Record<MongoLoggableComponent, SeverityLevel>;
+  maxDocumentLength: number;
+  logPath: Writable;
+
+  constructor(options: MongoLoggerOptions) {
+    // validate log path
+    if (typeof options.logPath === 'string') {
+      this.logPath =
+        options.logPath === 'stderr' || options.logPath === 'stdout'
+          ? process[options.logPath]
+          : fs.createWriteStream(options.logPath, { flags: 'a+' });
```

Comment:
I want team discussion around this but I think we should not initialize the write stream in the constructor.

- i/o in constructors is bad practice generally
- we have no mechanism to propagate an error (which occurs asynchronously) when opening the file.  regardless of how we handle errors later on, this is an error we will want to show users (we should at the very least tell them that the logger is turned off when they expect it turned on, even if we swallow up any other logging errors) but parseOptions is not async, so we can't propagate the error there and the mongo client isn't created yet so we can't emit an error event from parseOptions.

I think "initializing" the logger could be done in a separate step on first-operation (or connect, which is really the same thing) and then error handling is easy:

- We have the mongo client at that point, so we can emit an error event if we choose that route for error handling
- We are now in an async context, so we can throw an error asynchronously if we decide to go that route.

Thoughts @durran @dariakp @andymina @nbbeeken ?

Reply:
From slack: I don't feel strongly about the code's location and we have error handling as a follow up ticket, we could omit the createWriteStream case of the ternary altogether right now since it's unused here, and defer to the next bit of work to stick it in the right place and correctly link up / await outcomes from i/o as needed.